sdk/web-pubsub/web-pubsub-express/src/cloudEventsDispatcher.ts

```diff
@@ -0,0 +1,273 @@
+
+        if (receivedEvent.data) {
+          data = receivedEvent.data as string;
+          dataType = receivedEvent.datacontenttype === "application/json" ? "json" : "text";
+        } else if (receivedEvent.data_base64) {
+          data = Buffer.from(receivedEvent.data_base64, "base64");
+        } else {
+          throw new Error("empty data payload");
+        }
+
+        var userRequest: UserEventRequest = {
+          context: this.GetContext(receivedEvent, request.headers.host!),
+          data: data,
+          dataType: dataType
+        };
+        this.eventHandler.handleUserEvent!(userRequest, eventHandler);
+        return true;
+      }
+      default:
+        throw new Error(`Unknown EventType ${eventType}`);
+    }
+  }
+
+  private tryGetWebPubSubEvent(req: IncomingMessage): EventType | undefined {
+    // check ce-type to see if it is a valid WebPubSub CloudEvent request
+    const prefix = "azure.webpubsub.";
+    const connect = "azure.webpubsub.sys.connect";
+    const connected = "azure.webpubsub.sys.connected";
+    const disconnectd = "azure.webpubsub.sys.disconnected";
+    const userPrefix = "azure.webpubsub.user.";
+    var type = req.headers["ce-type"];
+    if (!type || typeof type !== "string" || !type.startsWith(prefix)) {
+      return undefined;
+    }
+    if (type.startsWith(userPrefix)) {
+      return EventType.UserEvent;
+    }
+    switch (type) {
+      case connect:
+        return EventType.Connect;
+
+      case connected:
+        return EventType.Connected;
+      case disconnectd:
+        return EventType.Disconnected;
+      default:
+        return undefined;
+    }
+  }
+
+  private GetContext(ce: CloudEvent, host: string): ConnectionContext {
+    var context = {
+      signature: ce["signature"] as string,
+      userId: ce["userid"] as string,
+      hub: ce["hub"] as string,
+      connectionId: ce["connectionid"] as string,
+      eventName: ce["eventname"] as string,
+      host: host
+    };
+
+    // TODO: validation
+    return context;
+  }
+
+  private async convertHttpToEvent(request: IncomingMessage): Promise<Message> {
+    const normalized: Message = {
+      headers: {},
+      body: ""
+    };
+    if (request.headers) {
+      for (const key in request.headers) {
+        if (Object.prototype.hasOwnProperty.call(request.headers, key)) {
+          const element = request.headers[key];
+          if (element === undefined) {
+            continue;
+          }
+          if (typeof element === "string") {
+            normalized.headers[key] = element;
+          } else {
+            normalized.headers[key] = element.join(",");
+          }
+        }
+      }
+    }
+
+    normalized.body = await this.readRequestBody(request);
+    return normalized;
+  }
+
+  private readRequestBody(req: IncomingMessage): Promise<string> {
+    return new Promise(function(resolve, reject) {
+      var body = "";
+      req.on("data", function(chunk) {
+        body += chunk;
```

Comment:
It looks like we're converting the chunk to a string then concatenating it. Can we instead store chunks as an array of buffers, then concatenate/stringify inside the `end` handler? I'm not sure what the incoming requests will look like but it'll help ensure we don't get any weird issues such as a UTF-8 character being on a boundary where it's partially sent in 2 chunks.

Reply:
nice catch